lib/internal/streams/web.js

```diff
@@ -0,0 +1,4643 @@
+    totalBytesToCopyRemaining -= bytesToCopy;
+  }
+
+  if (!ready) {
+    assert(!controller[kState].queueTotalSize);
+    assert(desc.bytesFilled > 0);
+    assert(desc.bytesFilled < elementSize);
+  }
+  return ready;
+}
+
+function readableByteStreamControllerProcessPullIntoDescriptorsUsingQueue(
+  controller) {
+  const {
+    closeRequested,
+    pendingPullIntos,
+    stream,
+  } = controller[kState];
+  assert(!closeRequested);
+  while (pendingPullIntos.length) {
+    if (!controller[kState].queueTotalSize)
+      return;
+    const desc = pendingPullIntos[0];
+    if (readableByteStreamControllerFillPullIntoDescriptorFromQueue(
+      controller,
+      desc)) {
+      readableByteStreamControllerShiftPendingPullInto(controller);
+      readableByteStreamControllerCommitPullIntoDescriptor(stream, desc);
+    }
+  }
+}
+
+function readableByteStreamControllerRespondInReadableState(
+  controller,
+  bytesWritten,
+  desc) {
+  const {
+    buffer,
+    bytesFilled,
+    byteLength,
+  } = desc;
+
+  if (bytesFilled + bytesWritten > byteLength)
+    throw new ERR_INVALID_STATE.RangeError('The buffer size is invalid');
+
+  readableByteStreamControllerFillHeadPullIntoDescriptor(
+    controller,
+    bytesWritten,
+    desc);
+
+  if (desc.bytesFilled < desc.elementSize)
+    return;
+
+  readableByteStreamControllerShiftPendingPullInto(controller);
+
+  const remainderSize = desc.bytesFilled % desc.elementSize;
+
+  if (remainderSize) {
+    const end = desc.byteOffset + desc.bytesFilled;
+    const start = end - remainderSize;
+    const remainder =
+      ArrayBufferPrototypeSlice(
+        buffer,
+        start,
+        end);
+    readableByteStreamControllerEnqueueChunkToQueue(
+      controller,
+      remainder,
+      0,
+      ArrayBufferGetByteLength(remainder));
+  }
+  desc.bytesFilled -= remainderSize;
+  readableByteStreamControllerCommitPullIntoDescriptor(
+    controller[kState].stream,
+    desc);
+  readableByteStreamControllerProcessPullIntoDescriptorsUsingQueue(controller);
+}
+
+function readableByteStreamControllerRespondWithNewView(controller, view) {
+  const {
+    stream,
+    pendingPullIntos,
+  } = controller[kState];
+  assert(pendingPullIntos.length);
+
+  const desc = pendingPullIntos[0];
+  assert(stream[kState].state !== 'errored');
+
+  if (!isArrayBufferView(view)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'view',
+      [
+        'Buffer',
+        'TypedArray',
+        'DataView',
+      ],
+      view);
+  }
+  const viewByteLength = ArrayBufferViewGetByteLength(view);
+  const viewByteOffset = ArrayBufferViewGetByteOffset(view);
+  const viewBuffer = ArrayBufferViewGetBuffer(view);
+
+  const {
+    byteOffset,
+    byteLength,
+    bytesFilled,
+  } = desc;
+
+  if (byteOffset + bytesFilled !== viewByteOffset)
+    throw new ERR_INVALID_ARG_VALUE.RangeError('view', view);
+
+  if (bytesFilled + viewByteOffset > byteLength)
+    throw new ERR_INVALID_ARG_VALUE.RangeError('view', view);
```

Comment:
This step seems to be missing:
> 8. If firstDescriptor’s buffer byte length is not view.[[ViewedArrayBuffer]].[[ByteLength]], throw a RangeError exception.

It seems you don't have the "buffer byte length" field in your pull-into descriptor representation either. It's a fairly recent addition, and its value is only needed for this one specific check. But still, it's an important check. 😁

Reply:
Ah right, I forgot to leave a todo comment here because I needed to check on this... The given view.byteLength does not have to match but it's underlying view.buffer.byteLength does? Is the assumption there that the view is most likely a slice of the byobRequest.view?

Reply:
Correct.

The intention is that you call `.respondWithNewView()` with a view whose buffer is either `=== byobRequest.view.buffer`, or is the "transferred version" of that buffer. This is necessary to uphold the contract of `byobReader.read(view)`: that promise must be resolved with a `{ done, value }` tuple whose `value.buffer` is the transferred version of `view.buffer`. The consumer must get their entire buffer back, since it may contain data from previous reads that they still want to use later.

The view's byte offset must be the same as `byobRequest.view.byteOffset`. So exactly like in `.respond(bytesWritten)`, the newly written bytes must be placed at the start of the BYOB request view.

The view's byte length must be at most `byobRequest.view.byteLength`, but it can be shorter. Once again, this is the same as with `.respond(bytesWritten)`, there `bytesWritten` must also be `<= byobRequest.view.byteLength`.

Effectively, these two calls are equivalent:
```javascript
byobRequest.respond(bytesWritten);
byobRequest.respondWithNewView(byobRequest.view.subarray(0, bytesWritten));
```
The advantage of the latter is that you can pass a view that is backed by a transferred version of `byobRequest.view.buffer`. This opens up more advanced use cases, for example you could transfer the BYOB request to a worker to fill it, then transfer it back and call `.respondWithNewView()`. This isn't possible with `.respond()`, since that requires that `byobRequest.view` **must not** have been transferred.

See [whatwg/streams#1114 (comment)](https://github.com/whatwg/streams/pull/1114#discussion_r600081235) and [#1123](https://github.com/whatwg/streams/pull/1123) for more context.

Reply:
Gotcha... makes perfect sense. Thank you! :-)